driver-core/src/main/com/mongodb/internal/TimeoutContext.java

```diff
@@ -125,8 +125,8 @@ public boolean hasTimeoutMS() {
      * @return true if the timeout has been set and it has expired
      */
     public boolean hasExpired() {
-        // Use timeout.remaining instead of timeout.hasExpired that measures in nanoseconds.
-        return timeout != null && !timeout.isInfinite() && timeout.remaining(MILLISECONDS) <= 0;
+        // TODO-CSOT remove this method (do not inline, but use lambdas)
+        return Timeout.run(timeout, NANOSECONDS, () -> false, (ns) -> false, () -> true, () -> false);
```

Comment:
I think changing to MS won't fully prevent a race, because for example if we check hasExpired at exactly 1ms (not expired) but a few nanoseconds elapse from then until the point we get remaining, remaining(ms) will round down returning 0. But I think the race is fixed, because since the `remaining` methods have been removed from prod, all the places where we could have gotten a 0 now explicitly check for expiry again and specify distinct expired/infinite branches based on only that single check.

Will comment on hasExpired below.